### PR TITLE
[8.x] Fix error handling

### LINKED ISSUE #40554: Memory leak in tests because HandleExceptions never cleans up
- Laravel Version: 8.x (though any version with the same logic has this issue)
- PHP Version: 8.0.6/8.1.1
- Database Driver & Version: -

### Description:
Theres a problem of [HandleExceptions](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php) keeping the handler itself and the old [Application](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Foundation/Application.php) instances throughout the whole test run, this is because of how `set_error_handler`, `set_exception_handler` and in addition `register_shutdown_function` work. All of these functions keep all the handlers in a memory stack, meaning they all reside in memory until the equivalent handler is restored or the application shuts down - which only happens when all tests have finished running.

To prove it, we can add:
```php
dump('cleanup', spl_object_id($this), spl_object_id($this->app));
```

to the the `__destruct`  handler or `handleShutdown` function of the `HandleExceptions` and it will only be executed after all tests are done. In addition, it will also execute n times, where n is the number of tests executed and all their ids are different meaning the handler and the application object it's holding to is no the same as previous execution, resulting in memory leak.

We can also test that this is the case using
```php
$prevHandler = set_error_handler([$this, 'handleError']);

// Make sure it's not the PHPUnit handler but our registered array
if (is_array($prevHandler)) {
    $isThis = $prevHandler[0] === $this; // always false
    $isSameApp = $prevHandler[0]->app === $this->app; // always false
}
```

### Steps To Reproduce:

While super basic and barebones test, I created 1000 tests with just bool assertion:

```php
protected function doAction()
{
    $this->assertTrue(true);
}

public function test_example()
{
    $this->doAction();
}

public function test_example2()
{
    $this->doAction();
}

// ... test_example1000
```

just commenting out or storing the handler as a static class variable to keep it single and just switching the $app variable in bootstrap
```php
set_error_handler([$this, 'handleError']);

set_exception_handler([$this, 'handleException']);

register_shutdown_function([$this, 'handleShutdown']);
```

saves 6 MB memory in total, tested with forced GC to avoid any memory fluctuations. Now if we have a `__destruct`, it will be called after each test and not after all tests are done, meaning it can now clean the object instantly without waiting for the process to finish. With more complex tests the memory difference can be greatly increased depending on the state the Application is holding on to.

I could make a PR but I have no idea what would be the best approach to keep a single exception handler instead of initiating it every time so knowledge about internals would help here. The main problem I see here is that if the HandleExceptions is to be instantiated every time, it needs to call `restore_error_handler` and `restore_exception_handler` before registering new handlers and `register_shutdown_function` should be registered only once no matter how many instances there are because it will prevent it from cleaning up.

### PR DESCRIPTION
**Pull request by X-Coder264**
Fixes problems described in https://github.com/laravel/framework/pull/40592#issuecomment-1021554516

TLDR; Laravel never cleaned up its handlers, resulting in the PHPUnit error handler never firing for unit tests that came after Laravel's integration tests, plus since the handlers were never cleaned it resulted in a memory leak.

In order to avoid having to call `$this->withoutDeprecationHandling()` in every Laravel integration test I've added that call in the `protected function deprecationHandling()` hook method which is now called from the base test class `setUp` method (users can override it in their base class if they want to change the behavior).

The changes made in the base test class here should also be ported over to Testbench, after which I can fix all the deprecation exceptions that will pop up on CI, before this PR can be merged.

Also note that this PR makes use of a terminating callback to restore the handlers, which also meant having to manually call `$this->app->terminate()` in the test teardown (it's weird that this was not already called as `terminate` is always called in a real HTTP request).

Fixes https://github.com/laravel/framework/issues/40554

### CODE REVIEW

File: src/Illuminate/Foundation/Testing/Concerns/InteractsWithDeprecationHandling.php [hunk 1]
```diff
@@ -42,6 +50,14 @@ protected function withoutDeprecationHandling()
                     throw new ErrorException($message, 0, $level, $file, $line);
                 }
             });
+
+            if (method_exists($this, 'beforeApplicationDestroyed')) {
+                $this->beforeApplicationDestroyed(function () {
+                    if (null !== $this->originalDeprecationHandler) {
+                        restore_error_handler();
```
Comment (donnysim) [inline src/Illuminate/Foundation/Testing/Concerns/InteractsWithDeprecationHandling.php:57]: We need to reset `$this->originalDeprecationHandler = null;` otherwise the memory leak persists because PHPUnit TestSuiteIterator creates a new class instance for each test method in the class and keeps everything in memory until test suite is done. After this I think it's looking good.

**Comment by donnysim** [comment]
There might still be something wrong, the deprecation errors are still not popping up in full test runs.

File: src/Illuminate/Foundation/Testing/Concerns/InteractsWithDeprecationHandling.php (hunk 1 above)
Comment (X-Coder264) [reply src/Illuminate/Foundation/Testing/Concerns/InteractsWithDeprecationHandling.php:58]: Done :+1:

**Comment by X-Coder264** [comment]
Yeah, the Testbench changes are needed before the deprecation exceptions pop up on CI.

Thank you for testing.

**Comment by nunomaduro** [comment]
Converting this to draft while I try to understand the reasoning behind this pull request.

**Comment by nunomaduro** [comment]
@X-Coder264 @donnysim Let me know what do you think about this alternative: https://github.com/laravel/framework/pull/40656.
